File: yc306/1623.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

// BIT stands for binary indexed tree.
type BIT []int

func newBIT(n int) BIT {
	return make([]int, n)
}

func (bit BIT) add(i, v int) {
	for i++; i <= len(bit); i += i & -i {
		bit[i-1] += v
	}
}

func (bit BIT) sum(i int) int {
	result := 0
	for i++; i > 0; i -= i & -i {
		result += bit[i-1]
	}
	return result
}

func (bit BIT) query(start int, stop int) int {
	return bit.sum(stop-1) - bit.sum(start-1)
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func main() {
	defer flush()

	n := readInt()
	r := make([]int, n)
	for i := 0; i < n; i++ {
		r[i] = readInt()
	}
	g := make([]int, n)
	for i := 0; i < n; i++ {
		g[i] = readInt()
	}
	b := make([]int, n)
	for i := 0; i < n; i++ {
		b[i] = readInt()
	}

	rr := make([]int, 3001)
	for i := 0; i < n; i++ {
		rr[r[i]]++
	}
	gg := make([]int, 3001)
	for i := 0; i < n; i++ {
		gg[g[i]]++
	}
	bb := make([]int, 3001)
	for i := 0; i < n; i++ {
		bb[b[i]]++
	}

	gb := make([][]int, 3001)
	for i := 0; i < 3001; i++ {
		if gg[i] == 0 {
			continue
		}
		for j := 0; j < 3001; j++ {
			if bb[j] == 0 {
				continue
			}
			gb[max(i, j)] = append(gb[max(i, j)], i<<12+j)
		}
	}

	result := 0
	bit := newBIT(6001)
	for i := 0; i < 3001; i++ {
		for _, v := range gb[i] {
			x := v >> 12
			y := v & 4095
			bit.add(x+y, gg[x]*bb[y])
		}
		result += rr[i] * bit.query(i+1, 6001)
	}
	println(result)
}

const (
	ioBufferSize = 1 * 1024 * 1024 // 1 MB
)

var stdinScanner = func() *bufio.Scanner {
	result := bufio.NewScanner(os.Stdin)
	result.Buffer(make([]byte, ioBufferSize), ioBufferSize)
	result.Split(bufio.ScanWords)
	return result
}()

func readString() string {
	stdinScanner.Scan()
	return stdinScanner.Text()
}

func readInt() int {
	result, err := strconv.Atoi(readString())
	if err != nil {
		panic(err)
	}
	return result
}

var stdoutWriter = bufio.NewWriter(os.Stdout)

func flush() {
	stdoutWriter.Flush()
}

func println(args ...interface{}) (int, error) {
	return fmt.Fprintln(stdoutWriter, args...)
}

File: yc306/1623.py
class BinaryIndexedTree:
    def __init__(self, size):
        self._data = [0] * size

    def add(self, i, x):
        data = self._data
        i += 1
        n = len(data)
        while i <= n:
            data[i - 1] += x
            i += i & -i

    def _sum(self, stop):
        data = self._data
        result = 0
        i = stop
        while i > 0:
            result += data[i - 1]
            i -= i & -i
        return result

    def range_sum(self, start, stop):
        return self._sum(stop) - self._sum(start)


n = int(input())
r = list(map(int, input().split()))
g = list(map(int, input().split()))
b = list(map(int, input().split()))

rr = [0] * 3001
for i in r:
    rr[i] += 1

gg = [0] * 3001
for i in g:
    gg[i] += 1

bb = [0] * 3001
for i in b:
    bb[i] += 1

gb = [[] for _ in range(3001)]
for i in range(1, 3001):
    x = gg[i]
    if x == 0:
        continue
    for j in range(1, 3001):
        y = bb[j]
        if y == 0:
            continue
        gb[max(i, j)].append((i << 12) + j)

bit = BinaryIndexedTree(6001)
result = 0
for i in range(1, 3001):
    for v in gb[i]:
        x = v >> 12
        y = v & 4095
        bit.add(x + y, gg[x] * bb[y])
    result += rr[i] * bit.range_sum(i + 1, 6001)
print(result)
